### PR TITLE
feat(ext/net): Refactor TCP socket listeners for future clustering mode

### PR DESCRIPTION
**Pull request by mmastrac**
Changes:

 - Implements a TCP socket listener that will allow for round-robin load-balancing in-process. 
 - Cleans up the raw networking code to make it easier to work with.

### CODE REVIEW

File: tests/unit/net_test.ts [hunk 1]
```diff
@@ -1204,33 +1203,22 @@ Deno.test({
       p1 = listener1.accept().then((conn) => {
         conn.close();
```
Comment (mmastrac) [inline tests/unit/net_test.ts:1204]: This was assuming that connections are always round-robin'd, but that's not a guarantee.

File: tests/unit/tls_test.ts [hunk 2]
```diff
@@ -1589,41 +1588,22 @@ Deno.test({
       p1 = listener1.accept().then((conn) => {
         conn.close();
         listener1Recv = true;
+        listener1.close();
```
Comment (mmastrac) [inline tests/unit/tls_test.ts:1591]: This was assuming that connections are always round-robin'd, but that's not a guarantee.

File: ext/net/tcp.rs [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+    socket2::Socket::new(Domain::IPV4, Type::STREAM, Some(Protocol::TCP))?
+  } else {
+    socket2::Socket::new(Domain::IPV6, Type::STREAM, Some(Protocol::TCP))?
+  };
+  #[cfg(target_os = "linux")]
+  if reuse_port {
+    socket.set_reuse_port(true)?;
+  }
+  #[cfg(not(windows))]
+  socket.set_reuse_address(true)?;
```
Comment (mmastrac) [inline ext/net/tcp.rs:162]: This existed before and prevents most of our tests from failing when we reuse a port binding in a subsequent test.

**Comment by ry** [comment]
SO_REUSEPORT is not the ideal way to load balance between threads. Each thread should eagerly accept connections on the same server socket.

There might be other reasons to support SO_REUSEPORT but I don’t agree with purpose set out in the description.

**Comment by mmastrac** [comment]
@ry I probably didn't describe it well -- it's more like emulation of SO_REUSEPORT.

I could also add a separate option where we attempt to share ports in-process, or make it a clustering option that chooses the best approach.

File: ext/net/ops.rs [hunk 4]
```diff
@@ -85,8 +86,7 @@ pub async fn op_net_accept_tcp(
   let resource = state
     .borrow()
     .resource_table
-    .get::<TcpListenerResource>(rid)
-    .map_err(|_| bad_resource("Listener has been closed"))?;
```
Comment (bartlomieju) [inline ext/net/ops.rs:89]: Can we keep this message? It's way more informative than just `BadResource`

File: ext/net/ops_tls.rs [hunk 5]
```diff
@@ -454,31 +454,14 @@ where
...
-  let std_listener: std::net::TcpListener = socket.into();
-  let tcp_listener = TcpListener::from_std(std_listener)?;
+
+  let tcp_listener = TcpListener::bind_direct(bind_addr, args.reuse_port)?;
   let local_addr = tcp_listener.local_addr()?;
 
-  let tls_listener_resource = TlsListenerResource {
-    tcp_listener: AsyncRefCell::new(tcp_listener),
-    tls_config: Arc::new(tls_config),
-    cancel_handle: Default::default(),
```
Comment (bartlomieju) [inline ext/net/ops_tls.rs:480]: Does removal of this handle here means that listening is not canceled when the resource is closed?

File: ext/net/raw.rs [hunk 6]
```diff
@@ -1,223 +1,322 @@
...
-    NetworkStream::Tls(value)
-  }
+/// A strongly-typed network listener resource for something that
+/// implements `NetworkListenerTrait`.
+pub struct NetworkListenerResource<T: NetworkStreamListenerTrait> {
+  pub listener: AsyncRefCell<T>,
+  /// Associated data for this resource. Not required.
+  #[allow(unused)]
+  pub data: T::ResourceData,
+  pub cancel: CancelHandle,
```
Comment (bartlomieju) [inline ext/net/raw.rs:42]: Ah, it seems it was hoisted here?

File: ext/net/tcp.rs [hunk 7]
```diff
@@ -0,0 +1,171 @@
...
+use std::collections::HashMap;
+use std::net::SocketAddr;
+use std::sync::Arc;
+
+use socket2::Domain;
+use socket2::Protocol;
+use socket2::Type;
+
+static CONNS: std::sync::OnceLock<std::sync::Mutex<Connections>> =
+  std::sync::OnceLock::new();
```
Comment (bartlomieju) [inline ext/net/tcp.rs:11]: Can you add a docstring here and explain purpose of this static?

File: ext/net/raw.rs (hunk 6 above)
Comment (mmastrac) [reply ext/net/raw.rs:42]: Correct!

**Pull request merged**